sql: Collapse sequential % characters in LIKE statements

ADustyOldMuffin opened the pull request:
Fixes #80192

util: Add a function for collapsing a certain character in strings as well as tests and benchmarks for the function. As mentioned above it will take "%%%hello%" -> given that string and '%' will produce "%hello%".

sql: Before the pattern passed to like_ops in both `colexeccmp` and `colexecsel` might contain repeating %, but these optimizer functions couldn't make use of them and the multiple "%%%" translate to just a single %. This change calls a utility function to cleanup the pattern string and collapse the % character to a single one if a repeating pattern of them are found.

Release note (performance optimization): SQL LIKE statements will now normalize statements including multiple "%" operators to a single character. E.g. "%foo%%%" -> "%foo%".

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=80432) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=80432) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=80432) it.</sub>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/80432)
<!-- Reviewable:end -->

nvanbenschoten commented:
Thanks for the contribution @ADustyOldMuffin! This looks like a nice change.

I see that `CollapseDupeChar` will reallocate and copy the input string even if no modification is needed. Can we try to avoid that so that the common case is cheaper? In other words, `CollapseDupeChar` would benefit from a fast-path that includes `return toCollapse`.

Also, we should wrap this utility in a benchmark. I suspect that if we do so, we'll find that copying byte-by-byte is less efficient than copying continuous chunks of the input string at a time.

ADustyOldMuffin commented:
Hello @nvanbenschoten and thanks for taking a look! I pushed some benchmarks and added a fast exit path for the common case. 

The results of the benchmarks on my machine were 

```
// Single
// goos: linux
// goarch: amd64
// pkg: github.com/cockroachdb/cockroach/pkg/util/strings
// cpu: Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
// BenchmarkSingleDupe-16    	20045674	        58.76 ns/op	       8 B/op	       1 allocs/op
// PASS
// ok  	github.com/cockroachdb/cockroach/pkg/util/strings	1.241s

// Multi
// goos: linux
// goarch: amd64
// pkg: github.com/cockroachdb/cockroach/pkg/util/strings
// cpu: Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
// BenchmarkMultipleDupe-16    	19472762	        61.68 ns/op	       8 B/op	       1 allocs/op
// PASS
// ok  	github.com/cockroachdb/cockroach/pkg/util/strings	1.266s

// Spaced
// goos: linux
// goarch: amd64
// pkg: github.com/cockroachdb/cockroach/pkg/util/strings
// cpu: Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
// BenchmarkSpacedDupe-16    	11816037	        99.78 ns/op	      24 B/op	       2 allocs/op
// PASS
// ok  	github.com/cockroachdb/cockroach/pkg/util/strings	1.284s
```

I Then changed the main loop to be this so it'd be writing chunks instead of every run.
```
	var builder strings.Builder
	hadPrevious := false
	begIndx := 0
	endIndx := 0
	for _, r := range toCollapse {
		if r != collapseChar {
			hadPrevious = false
			endIndx++
			continue
		}

		// Our character matches
		if hadPrevious && begIndx != endIndx {
			builder.WriteString(toCollapse[begIndx:endIndx])
			endIndx++
			begIndx = endIndx
			continue
		} else if hadPrevious {
			endIndx++
			begIndx = endIndx
			continue
		}

		hadPrevious = true
		endIndx++
	}

	builder.WriteString(toCollapse[begIndx:endIndx])

	return builder.String()
```

The results of the benchmarks were

```
// Single
// goos: linux
// goarch: amd64
// pkg: github.com/cockroachdb/cockroach/pkg/util/strings
// cpu: Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
// BenchmarkSingleDupe-16    	23591236	        50.77 ns/op	       8 B/op	       1 allocs/op
// PASS
// ok  	github.com/cockroachdb/cockroach/pkg/util/strings	1.252s

// Multi
// goos: linux
// goarch: amd64
// pkg: github.com/cockroachdb/cockroach/pkg/util/strings
// cpu: Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
// BenchmarkMultipleDupe-16    	18958507	        62.69 ns/op	       8 B/op	       1 allocs/op
// PASS
// ok  	github.com/cockroachdb/cockroach/pkg/util/strings	1.255s

// Spaced
// goos: linux
// goarch: amd64
// pkg: github.com/cockroachdb/cockroach/pkg/util/strings
// cpu: Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
// BenchmarkSpacedDupe-16    	12362629	       103.0 ns/op	      24 B/op	       2 allocs/op
// PASS
// ok  	github.com/cockroachdb/cockroach/pkg/util/strings	1.374s
```

So with my attempt, while writing chunks is better for a simple single dupe, it gets worse as more dupes are present, and with a gap in the dupes `%%as%%an%%%example%%%` it's actually approaching a significant difference.

ADustyOldMuffin commented:
So my attempt at writing chunks IMO is worse than eating the 8ns for the simple case, and so I didn't commit it. I'm open to hearing a better way to write chunks though and testing it out.

ADustyOldMuffin commented:
To me the results seem fairly similar, but I'm also sure that my initial attempt at writing chunks is inefficient so feel free to correct me on it!

rytaft reviewed (commented):
Thanks for the contribution, @adustyoldmuffin!

I'll let @nvanbenschoten respond regarding the benchmark results, however I'm guessing that you will need to run them multiple times to get consistent results. @nvanbenschoten created a useful tool for doing this and comparing the results: https://github.com/nvanbenschoten/benchdiff

Please squash your commits to correspond to one or more logical changes. I think one commit should suffice for this change. For more info, see this wiki page: https://wiki.crdb.io/wiki/spaces/CRDB/pages/1411744698/Decomposing+changes+and+updating+PRs+during+reviews

Also some nits about the PR message:
- Update the PR title to include the affected packages, e.g.: "sql, util: remove sequential % in LIKE operators"
- Typically we put "Fixes \<issue number\>" on its own line
- The PR message should be mostly self-contained, so readers don't need to look at the issue or the code changes to understand the message. Typically we just copy the text from all commit messages into the PR message.

> Reviewed 3 of 3 files at r1, 4 of 4 files at r2, 1 of 1 files at r3, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80432#-:-N0RXfVX5uFDD1t14T_g:b-5dndmv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ADustyOldMuffin)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80432#-N0RV83X3ZqMXUWlLTbc):*
nit: commit messages and PR messages should start with the affected package. e.g., `util: add utility function for cleaning strings`

For more details: https://wiki.crdb.io/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages
___
*[`pkg/util/strings/string_util.go` line 26 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/80432#-N0RYSUr7TrBCncRUH70:-N0RYSUr7TrBCncRUH71:b1k99d1) ([raw file](https://github.com/cockroachdb/cockroach/blob/be9aa7b66dc1b66533e6154f73d7023dbb0eb4f9/pkg/util/strings/string_util.go#L26)):*
> ```Go
> // single instance.
> // E.g. CollapseDupeChar("wwwhy hello there", 'w') returns "why hello there"
> func CollapseDupeChar(toCollapse string, collapseChar rune) string {
> ```

I think this function could go in the existing file https://github.com/cockroachdb/cockroach/blob/1c9731ce1d3454f95933efcad1bf5003dddc1604/pkg/util/strings.go, and the test could go in https://github.com/cockroachdb/cockroach/blob/1c9731ce1d3454f95933efcad1bf5003dddc1604/pkg/util/strings_test.go

Then you wouldn't need to create this new package or file.
___
*[`pkg/util/strings/string_util.go` line 39 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/80432#-N0RsOsmCk0MRyqo3-V2:-N0RsOsmCk0MRyqo3-V3:bxjusch) ([raw file](https://github.com/cockroachdb/cockroach/blob/1850fd1d194f4101925ecee43d90158fee665b1b/pkg/util/strings/string_util.go#L39)):*
> ```Go
> 	begIndx := 0
> 	endIndx := 0
> 	for _, r := range toCollapse {
> ```

I feel like this logic would be clearer if you actually use the current index from the for loop instead of incrementing `endIdx`, e.g.:

`for i, r := range toCollapse {`

Then remove all instances of `endIndx++` and replace other uses of `endIdx` with `i`

Also looks like you're no longer using `r`, so you can simplify to `for i := range toCollapse {`
___
*[`pkg/util/strings/string_util_test.go` line 74 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/80432#-N0RmcPE9qaR0TBuNXwu:-N0RmcPE9qaR0TBuNXwv:bpb9zom) ([raw file](https://github.com/cockroachdb/cockroach/blob/1850fd1d194f4101925ecee43d90158fee665b1b/pkg/util/strings/string_util_test.go#L74)):*
> ```Go
> 		CollapseDupeChar(toTest, '%')
> 	}
> }
> ```

It would be best if you could combine these all into a single test, `BenchmarkCollapseDupeChar`, with different cases. For example, notice how we're using `b.Run` with a different name for each case here: https://github.com/cockroachdb/cockroach/blob/0860b7fe03d1df7524f1b23e86a8912aa329e464/pkg/bench/bench_test.go#L379-L404

and here:
https://github.com/cockroachdb/cockroach/blob/0860b7fe03d1df7524f1b23e86a8912aa329e464/pkg/bench/bench_test.go#L408-L536


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80432)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ADustyOldMuffin)

___
*[`pkg/util/strings/string_util.go` line 31 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/80432#-N0W7JSi-9mqYiUXSL9L:-N0W7JSi-9mqYiUXSL9M:b-y355uj) ([raw file](https://github.com/cockroachdb/cockroach/blob/1850fd1d194f4101925ecee43d90158fee665b1b/pkg/util/strings/string_util.go#L31)):*
> ```Go
> 
> 	// There are no repeating characters
> 	if strings.Index(toCollapse, string([]rune{collapseChar, collapseChar})) == -1 {
> ```

This works, but it's expensive, as it requires a second pass through the input string and may also (depending on escape analysis) require a heap allocation of the search slice. Part of why we may want to copy in chunks is that we can avoid the second pass. Instead, we can defer this fast-path for the end of the function with a check like:
```go
if begIndx == 0 {
    return toCollapse
}
```

Also, in your example, do we need to manipulate the `endIndx ` variable? Isn't that just `for endIndx, r := range toCollapse {`?


<!-- Sent from Reviewable.io -->

ADustyOldMuffin committed:
sql: Collapse sequential % characters in LIKE statements

util: Add a function for collapsing a certain character in strings as well as tests and benchmarks for the function. As mentioned above it will take "%%%hello%" -> given that string and '%' will produce "%hello%".

sql: Before the pattern passed to like_ops in both `colexeccmp` and `colexecsel` might contain repeating %, but these optimizer functions couldn't make use of them and the multiple "%%%" translate to just a single %. This change calls a utility function to cleanup the pattern string and collapse the % character to a single one if a repeating pattern of them are found.

Release note (performance optimization): SQL LIKE statements will now normalize statements including multiple "%" operators to a single character. E.g. "%foo%%%" -> "%foo%".

ADustyOldMuffin commented:
@rytaft 👋🏻 I cleaned up the commits and the commit message as requested, please let me know if any of that is incorrect still.

Thanks for pointing out the util! I looked for a string utility but was foolishly looking for a folder not a file!

I also modified the logic as you suggested as well, doing away with `endIndx` and only having a single write path in the loop. 

@nvanbenschoten Thanks for the clarification and advise! Using your suggestions I move the check to after the logic, and the need to write at all if the string is fine.

Benchmarks with the current setup look like 
```
goos: linux
goarch: amd64
pkg: github.com/cockroachdb/cockroach/pkg/util
cpu: Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
BenchmarkCollapseDupeChar/NoDupe/count=1-16 	        124362252	        9.506 ns/op	       0 B/op	       0 allocs/op
BenchmarkCollapseDupeChar/NoDupe/count=10-16         	125839725	        9.528 ns/op	       0 B/op	       0 allocs/op
BenchmarkCollapseDupeChar/NoDupe/count=100-16        	126349200	        9.508 ns/op	       0 B/op	       0 allocs/op
BenchmarkCollapseDupeChar/NoDupe/count=1000-16       	125868572	        9.826 ns/op	       0 B/op	       0 allocs/op
BenchmarkCollapseDupeChar/SingleDupe/count=1-16      	36156650	        32.36 ns/op	       8 B/op	       1 allocs/op
BenchmarkCollapseDupeChar/SingleDupe/count=10-16     	35547456	        32.05 ns/op	       8 B/op	       1 allocs/op
BenchmarkCollapseDupeChar/SingleDupe/count=100-16    	36235418	        32.06 ns/op	       8 B/op	       1 allocs/op
BenchmarkCollapseDupeChar/SingleDupe/count=1000-16   	36647972	        32.12 ns/op	       8 B/op	       1 allocs/op
BenchmarkCollapseDupeChar/MultipleDupe/count=1-16    	28418353	        41.73 ns/op	       8 B/op	       1 allocs/op
BenchmarkCollapseDupeChar/MultipleDupe/count=10-16   	28371387	        41.70 ns/op	       8 B/op	       1 allocs/op
BenchmarkCollapseDupeChar/MultipleDupe/count=100-16  	28360509	        41.88 ns/op	       8 B/op	       1 allocs/op
BenchmarkCollapseDupeChar/MultipleDupe/count=1000-16 	28319415	        41.68 ns/op	       8 B/op	       1 allocs/op
BenchmarkCollapseDupeChar/SpacedDupe/count=1-16      	15690753	        75.60 ns/op	      24 B/op	       2 allocs/op
BenchmarkCollapseDupeChar/SpacedDupe/count=10-16     	15750147	        75.69 ns/op	      24 B/op	       2 allocs/op
BenchmarkCollapseDupeChar/SpacedDupe/count=100-16    	15767277	        75.77 ns/op	      24 B/op	       2 allocs/op
BenchmarkCollapseDupeChar/SpacedDupe/count=1000-16   	15713727	        75.50 ns/op	      24 B/op	       2 allocs/op
```

mgartner commented:
@rytaft Do you think this should be implemented as a normalization rule instead of in the vectorized engine?

rytaft reviewed (commented):
Looks much better @adustyoldmuffin, thanks! Just a couple more nits with the commit message below. Also don't forget to update the PR title.

> @rytaft Do you think this should be implemented as a normalization rule instead of in the vectorized engine?

Yes good point, @mgartner  -- looks like the issue you opened was for the purpose of improving optimization, so this would need to happen earlier.

> Reviewed 9 of 9 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80432#-:-N0XliYTAICN_4ucukGj:bd4462n)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ADustyOldMuffin)

___
*[`-- commits` line 2 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/80432#-N0Xm9Wf4cqhJwdFN0Dg):*
This is the line that should be updated to include the package names.
___
*[`-- commits` line 4 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/80432#-N0XlrkB7bFSLoVGBFSe):*
I think this would fit the release note category of "performance optimization".

Also, the release note should be the last thing in the commit message. The text you've added below isn't really user facing, so it should go above the release note.


<!-- Sent from Reviewable.io -->
